### ISSUE TITLE
Split and Concat layers

### ISSUE THREAD

**Issue by classner** [tiny-dnn/tiny-dnn#146]
Hi @nyanp,

great work with tiny-cnn! It should be super helpful for CNN deployment!

I was just testing its abilities and tried a GoogLeNet architecture, when I noticed that it does not support concat/split layers (error: tiny-cnn does not support this layer type:Split). This should be quite easy to implement and would extend its 'range' quite a bit with networks requiring this capability.

**Comment by jamseams** [comment]
Hey @classner,

I saw your post about testing the GoogleNet arch. so I'd assume you used the caffe_converter? I had some trouble running it with the bvlc_alexnet and the bvlc_reference_caffenet pretrained models.

I am curious if you also ran into this issue:  https://github.com/nyanp/tiny-cnn/issues/143

I'd really appreciate your input.

Servus!

Juan

**Comment by classner** [comment]
Hi!

No, that issue (#143) did not come up yet, but I guess that could still come up after network loading finished. If I interpret the output right, for me tiny-cnn did not even complete model loading and construction. Could you construct the nets without pushing anything through (i.e., run `    reload_weight_from_caffe_protobinary(trained_file, net.get());` successfully)?

**Comment by jamseams** [comment]
Hey,

Yes, my reload_weight_from_caffe_protobinary(trained_file, net.get()); was successful.

DId you run into the problem of input_param not recognized? If so, just checkout a previous commit of the deploy.prototxt file of the pretrained model in caffe.

Then after that your program should continue succesfully. (and we could see if you run into my same issue.)

Let me know if I could be of further help. =)

Juan

**Comment by classner** [comment]
Hi @nyanp ,

any news on this matter? The split layer only forwards the data to two outputs. I could probably even implement it myself, with a pointer in the right direction. Could you give me some pointers here?

**Comment by nyanp** [comment]
@classner 
Hi, thank you for nice suggestion! Currently we are changing an architecture of tiny-cnn in feat/generic-computational-graph branch to handle generic (non-sequential) models. After that, we can handle split/concat layer and GoogleNet-like networks. We already have [experimental implementation of concat layer](https://github.com/nyanp/tiny-cnn/blob/feat/generic-computational-graph/tiny_cnn/layers/concat_layer.h), and [guide to write a new layer](https://github.com/nyanp/tiny-cnn/blob/feat/generic-computational-graph/doc/Adding%20a%20new%20layer.md). It will be merged and released soon. stay tuned :)

**Comment by classner** [comment]
Sounds great! :)

**Comment by classner** [comment]
Hi @nyanp,

any news regarding concat and split?

**Comment by nyanp** [comment]
@classner 
From v0.1.0, we are able to handle multi-input / muti-output networks like googlenet :) Currently caffe-converter doesn't support googlenet yet, but we'll solve it in the near release (at least in the end of august).

**Comment by badlooop** [comment]
Hi @nyanp 
Any update?

I am trying to convert a pre-trained SqueezeNet model from caffe to tiny-dnn. If the converter works, it will be much helpful.

Many thanks!

**Comment by shenyuhit** [comment]
@nyanp  
Any update?

I am trying to convert a pre-trained SqueezeNet model from caffe to tiny-dnn. If the converter works, it will be much helpful.

Many thanks!

**Comment by maybepossible** [comment]
@nyanp
 Hi, does converter support googlenet now? when I try the google net, it always says _unsupported padding type_ at the beginning of the first branch.

**Comment by bhack** [comment]
You are entering in one or both of this two conditions https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/io/caffe/layer_factory_impl.h#L268,L284?

**Comment by reunanen** [comment]
> when I try the google net, it always says unsupported padding type at the beginning of the first branch.

@maybepossible, I think I was able to circumvent this issue (though I did not test everything thoroughly). See my comment [here](https://github.com/tiny-dnn/tiny-dnn/issues/459#issuecomment-267341135) – in particular, the link to the commit.

**Comment by bhack** [comment]
Can anyone retry on master? We have merged https://github.com/tiny-dnn/tiny-dnn/pull/557

**Comment by Randl** [comment]
@classner @maybepossible @jamseams Do you still have any issues?

**Comment by tomjaguarpaw** [comment]
@randi Split and Concat are still not supported, so yes, I would say there are still issues!

**Comment by ChenZhaobin** [comment]
Hi @nyanp
Any update?

I am trying to convert a pre-trained SqueezeNet model from caffe to tiny-dnn. If the converter works, it will be much helpful.

Many thanks!
